File: sentinel/engine-ts/src/m1_ingest.ts
/** M1 — Ingest & Validate. Quarantines (never drops silently) anything
 * structurally invalid, physically impossible, or outside the controlled
 * vocabularies. Bounds and unit conversions are content, not code.
 */
import { q6 } from "./canonical.js";
import { NUMERIC_TYPES, OBSERVATION_TYPES } from "./constants.js";
import type { ContentHandle } from "./content.js";
import { parseTs } from "./timeutil.js";

export const VALID_SOURCES = ["wearable_sensor", "inline_gauge", "manual_entry", "fixed_monitor", "event_report"];

export interface Quarantined {
  obs_id: string;
  reason: string;
  // Present only for out_of_bounds records so M8 can still test never-ignore
  // bounds against the rejected value (DECISIONS.md D13.5).
  type?: string;
  value?: unknown;
}

export interface NormalizedObs {
  obs_id: string;
  ts: number;
  timestamp: string;
  type: string;
  source: string;
  quality_meta: any;
  text?: string;
  event_id?: string;
  value?: any;
  unit?: string;
  // M2 annotations:
  quality?: string;
  rules_fired?: string[];
  mechanism_protected?: boolean;
}

function quarantine(quarantined: Quarantined[], obs: any, reason: string): void {
  const obsId = typeof obs === "object" && obs !== null && !Array.isArray(obs) ? obs.obs_id : undefined;
  quarantined.push({ obs_id: typeof obsId === "string" ? obsId : "(missing)", reason });
}

function isPlainObject(v: unknown): boolean {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/** Python `x or {}` for quality_meta. */
function orEmpty(v: any): any {
  if (v === null || v === undefined || v === false || v === "" || v === 0) {
    return {};
  }
  if (Array.isArray(v) && v.length === 0) {
    return {};
  }
  if (isPlainObject(v) && Object.keys(v).length === 0) {
    return {};
  }
  return v;
}

/** Returns [accepted, quarantined, notes]. Accepted observations are
 * normalized: canonical units, epoch-second `ts`, sorted by (ts, obs_id). */
export function ingest(
  observations: any[],
  content: ContentHandle,
  flags: Set<string>,
): [NormalizedObs[], Quarantined[], NormalizedObs[]] {
  const bounds = content.tables["operational_bounds"].bounds;
  const enums = content.tables["operational_bounds"].enums;
  const unitsTable = content.tables["units"];
  const eventIds = new Set<string>(content.tables["event_codes"].codes.map((c: any) => c.event_id));
  const conversions = new Map<string, any>();
  for (const c of unitsTable.conversions) {
    conversions.set(`${c.type}\u0000${c.from}`, c);
  }

  const accepted: NormalizedObs[] = [];
  const quarantined: Quarantined[] = [];
  const notes: NormalizedObs[] = [];
  const seenIds = new Set<string>();

  for (const obs of observations) {
    if (!isPlainObject(obs)) {
      quarantine(quarantined, {}, "invalid_structure");
      continue;
    }
    const obsId = obs.obs_id;
    if (typeof obsId !== "string" || obsId === "") {
      quarantine(quarantined, obs, "invalid_structure:obs_id");
      continue;
    }
    if (seenIds.has(obsId)) {
      quarantine(quarantined, obs, "duplicate_obs_id");
      flags.add("duplicate_obs_id");
      continue;
    }
    const otype = obs.type;
    if (typeof otype !== "string" || !OBSERVATION_TYPES.includes(otype)) {
      quarantine(quarantined, obs, `unknown_type:${pyStrOpt(otype)}`);
      continue;
    }
    const source = obs.source;
    if (typeof source !== "string" || !VALID_SOURCES.includes(source)) {
      quarantine(quarantined, obs, `unknown_source:${pyStrOpt(source)}`);
      continue;
    }
    let ts: number;
    try {
      ts = parseTs(obs.timestamp);
    } catch {
      quarantine(quarantined, obs, "invalid_timestamp");
      continue;
    }

    const norm: NormalizedObs = {
      obs_id: obsId,
      ts,
      timestamp: obs.timestamp,
      type: otype,
      source,
      quality_meta: orEmpty(obs.quality_meta),
    };

    if (otype === "free_text_note") {
      // Logged, NEVER parsed for logic (§2).
      norm.text = typeof obs.text === "string" ? obs.text : "";
      seenIds.add(obsId);
      notes.push(norm);
      continue;
    }

    if (otype === "event") {
      const eventId = obs.event_id;
      if (typeof eventId !== "string") {
        quarantine(quarantined, obs, "invalid_value:event_id_missing");
        continue;
      }
      if (!eventIds.has(eventId)) {
        quarantine(quarantined, obs, `unknown_event_code:${eventId}`);
        flags.add("unknown_event_code");
        continue;
      }
      norm.event_id = eventId;
      seenIds.add(obsId);
      accepted.push(norm);
      continue;
    }

    let value = obs.value;
    if (otype in enums) {
      if (!enums[otype].includes(value)) {
        quarantine(quarantined, obs, `invalid_value:${pyStrOpt(value)}`);
        continue;
      }
      norm.value = value;
    } else if (NUMERIC_TYPES.includes(otype)) {
      if (typeof value !== "number") {
        quarantine(quarantined, obs, "invalid_value:not_numeric");
        continue;
      }
      const unit = obs.unit;
      const canonicalUnit = unitsTable.canonical[otype];
      if (unit !== null && unit !== undefined && unit !== canonicalUnit) {
        const conv = conversions.get(`${otype}\u0000${unit}`);
        if (conv === undefined) {
          quarantine(quarantined, obs, `unknown_unit:${pyStrOpt(unit)}`);
          continue;
        }
        value = q6(value * conv.factor + conv.offset);
      }
      const b = bounds[otype];
      if (b === undefined || value < b.min || value > b.max) {
        quarantined.push({ obs_id: obsId, reason: `out_of_bounds:${otype}`, type: otype, value });
        flags.add("data_rejected");
        continue;
      }
      norm.value = value;
      norm.unit = canonicalUnit;
    }
    seenIds.add(obsId);
    accepted.push(norm);
  }

  if (quarantined.some((q) => q.reason.startsWith("out_of_bounds") || q.reason.startsWith("invalid_") || q.reason.startsWith("unknown_"))) {
    flags.add("data_rejected");
  }

  const byTsThenId = (a: NormalizedObs, b: NormalizedObs) =>
    a.ts - b.ts || (a.obs_id < b.obs_id ? -1 : a.obs_id > b.obs_id ? 1 : 0);
  accepted.sort(byTsThenId);
  notes.sort(byTsThenId);
  return [accepted, quarantined, notes];
}

/** Python str() of an arbitrary quarantine-reason value (None → "None"). */
function pyStrOpt(v: unknown): string {
  if (v === null || v === undefined) return "None";
  if (v === true) return "True";
  if (v === false) return "False";
  return String(v);
}

File: sentinel/engine-py/sentinel/m1_ingest.py
"""M1 — Ingest & Validate. Quarantines (never drops silently) anything
structurally invalid, physically impossible, or outside the controlled
vocabularies. Bounds and unit conversions are content, not code.
"""
from __future__ import annotations

from .canonical import fmt_val, q6
from .constants import NUMERIC_TYPES, OBSERVATION_TYPES
from .timeutil import parse_ts

VALID_SOURCES = ["wearable_sensor", "inline_gauge", "manual_entry", "fixed_monitor", "event_report"]


def _quarantine(quarantined, obs, reason):
    obs_id = obs.get("obs_id") if isinstance(obs, dict) else None
    quarantined.append({"obs_id": obs_id if isinstance(obs_id, str) else "(missing)", "reason": reason})


def ingest(observations, content, flags):
    """Returns (accepted, quarantined, notes). Accepted observations are
    normalized: canonical units, epoch-second `ts`, sorted by (ts, obs_id)."""
    bounds = content["tables"]["operational_bounds"]["bounds"]
    enums = content["tables"]["operational_bounds"]["enums"]
    units_table = content["tables"]["units"]
    event_ids = {c["event_id"] for c in content["tables"]["event_codes"]["codes"]}
    conversions = {(c["type"], c["from"]): c for c in units_table["conversions"]}

    accepted, quarantined, notes = [], [], []
    seen_ids = set()

    for obs in observations:
        if not isinstance(obs, dict):
            _quarantine(quarantined, {}, "invalid_structure")
            continue
        obs_id = obs.get("obs_id")
        if not isinstance(obs_id, str) or not obs_id:
            _quarantine(quarantined, obs, "invalid_structure:obs_id")
            continue
        if not obs_id.isascii():
            # ASCII-only ids keep the (timestamp, obs_id) tie-break sort and
            # canonical hashing byte-identical across runtimes.
            _quarantine(quarantined, obs, "invalid_structure:obs_id_not_ascii")
            continue
        if obs_id in seen_ids:
            _quarantine(quarantined, obs, "duplicate_obs_id")
            flags.add("duplicate_obs_id")
            continue
        otype = obs.get("type")
        if otype not in OBSERVATION_TYPES:
            _quarantine(quarantined, obs, f"unknown_type:{otype}")
            continue
        source = obs.get("source")
        if source not in VALID_SOURCES:
            _quarantine(quarantined, obs, f"unknown_source:{source}")
            continue
        try:
            ts = parse_ts(obs.get("timestamp"))
        except ValueError:
            _quarantine(quarantined, obs, "invalid_timestamp")
            continue

        stream_id = obs.get("stream_id")
        if stream_id is not None:
            if not isinstance(stream_id, str) or not stream_id or not stream_id.isascii():
                _quarantine(quarantined, obs, "invalid_structure:stream_id")
                continue

        norm = {
            "obs_id": obs_id,
            "ts": ts,
            "timestamp": obs["timestamp"],
            "type": otype,
            "source": source,
            # Stream identity: device/channel instance if given, else the
            # source class. All per-stream screening keys off this (D19).
            "stream": stream_id if stream_id is not None else source,
            "quality_meta": obs.get("quality_meta") or {},
        }

        if otype == "free_text_note":
            # Logged, NEVER parsed for logic (§2).
            norm["text"] = obs.get("text") if isinstance(obs.get("text"), str) else ""
            seen_ids.add(obs_id)
            notes.append(norm)
            continue

        if otype == "event":
            event_id = obs.get("event_id")
            if not isinstance(event_id, str):
                _quarantine(quarantined, obs, "invalid_value:event_id_missing")
                continue
            if event_id not in event_ids:
                _quarantine(quarantined, obs, f"unknown_event_code:{event_id}")
                flags.add("unknown_event_code")
                continue
            norm["event_id"] = event_id
            seen_ids.add(obs_id)
            accepted.append(norm)
            continue

        value = obs.get("value")
        if otype in enums:
            if value not in enums[otype]:
                # fmt_val keeps numeric interpolation byte-identical with
                # engine-ts (Python str(3.0) is "3.0"; JS String(3.0) is "3").
                _quarantine(quarantined, obs, f"invalid_value:{fmt_val(value)}")
                continue
            norm["value"] = value
        elif otype in NUMERIC_TYPES:
            if isinstance(value, bool) or not isinstance(value, (int, float)):
                _quarantine(quarantined, obs, "invalid_value:not_numeric")
                continue
            if value != value or value in (float("inf"), float("-inf")):
                # NaN passes < / > bounds comparisons; it must never reach
                # feature math or the audit log.
                _quarantine(quarantined, obs, "invalid_value:not_finite")
                flags.add("data_rejected")
                continue
            unit = obs.get("unit")
            canonical_unit = units_table["canonical"].get(otype)
            if unit is not None and unit != canonical_unit:
                conv = conversions.get((otype, unit))
                if conv is None:
                    _quarantine(quarantined, obs, f"unknown_unit:{unit}")
                    continue
                value = q6(value * conv["factor"] + conv["offset"])
            b = bounds.get(otype)
            if b is None or value < b["min"] or value > b["max"]:
                # Value is kept on the quarantine record so M8 can still test
                # never-ignore bounds against it (extreme-but-real vs garbage
                # is undecidable here; either way it must not vanish).
                quarantined.append({"obs_id": obs_id, "reason": f"out_of_bounds:{otype}",
                                    "type": otype, "value": value})
                flags.add("data_rejected")
                continue
            norm["value"] = value
            norm["unit"] = canonical_unit
        seen_ids.add(obs_id)
        accepted.append(norm)

    if any(q["reason"].startswith(("out_of_bounds", "invalid_", "unknown_")) for q in quarantined):
        flags.add("data_rejected")

    accepted.sort(key=lambda o: (o["ts"], o["obs_id"]))
    notes.sort(key=lambda o: (o["ts"], o["obs_id"]))
    return accepted, quarantined, notes
